services/src/main/java/org/keycloak/social/apple/AppleIdentityProvider.java

```diff
@@ -0,0 +1,311 @@
+import org.keycloak.broker.social.SocialIdentityProvider;
+import org.keycloak.common.ClientConnection;
+import org.keycloak.common.util.Time;
+import org.keycloak.crypto.*;
+import org.keycloak.events.Errors;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.events.EventType;
+import org.keycloak.jose.jws.JWSBuilder;
+import org.keycloak.jose.jws.JWSInput;
+import org.keycloak.jose.jws.JWSInputException;
+import org.keycloak.jose.jws.crypto.RSAProvider;
+import org.keycloak.keys.loader.PublicKeyStorageManager;
+import org.keycloak.models.*;
+import org.keycloak.protocol.oidc.OIDCLoginProtocol;
+import org.keycloak.representations.JsonWebToken;
+import org.keycloak.services.ErrorPage;
+import org.keycloak.services.Urls;
+import org.keycloak.services.messages.Messages;
+
+import javax.ws.rs.FormParam;
+import javax.ws.rs.POST;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.*;
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.util.Base64;
+
+/**
+ * @author Emilien Bondu
+ */
+public class AppleIdentityProvider extends AbstractOAuth2IdentityProvider<AppleIdentityProviderConfig> implements
+        SocialIdentityProvider<AppleIdentityProviderConfig> {
+
+    private static final String OAUTH2_PARAMETER_CODE = "code";
+
+    private static final String OAUTH2_PARAMETER_STATE = "state";
+
+    private static final String OAUTH2_PARAMETER_USER = "user";
+
+    public static final String ACCESS_DENIED = "access_denied";
+
+    protected static ObjectMapper mapper = new ObjectMapper();
+
+    public static final String AUTH_URL = "https://appleid.apple.com/auth/authorize?response_mode=form_post";
+
+    public static final String TOKEN_URL = "https://appleid.apple.com/auth/token";
+
+    public static final String ISSUER = "https://appleid.apple.com";
+
+    public static final String JWKS_URL = "https://appleid.apple.com/auth/keys";
+
+    public static final String EMAIL_SCOPE = "email";
+
+    public static final String NAME_SCOPE = "name";
+
+    protected static final Logger logger = Logger.getLogger(AppleIdentityProvider.class);
+
+    public AppleIdentityProvider(KeycloakSession session, AppleIdentityProviderConfig config) {
+        super(session, config);
+
+        config.setAuthorizationUrl(AUTH_URL);
+        config.setTokenUrl(TOKEN_URL);
+        config.setClientAuthMethod(OIDCLoginProtocol.CLIENT_SECRET_POST);
+        config.setIssuer(ISSUER);
+        config.setUseJwksUrl(true);
+        config.setValidateSignature(true);
+        config.setJwksUrl(JWKS_URL);
+        String defaultScope = config.getDefaultScope();
+
+        String privateKeyContent = config.getP8Content();
+
+        try {
+            KeyFactory kf = KeyFactory.getInstance("ECDSA");
```

Comment:
I think we should probably cache the private key instance and avoid creating it all the time.

I'm also wondering if we should probably re-use realm keys for importing this key and then just reference it when configuring the provider. But I think that would involve other things such as support for ECDSA.

Another consideration is that maybe you can store the JWS just once when creating a new apple broker so that you just obtain the secret when actually authenticating.

Reply:
You are right regarding the private key / secret caching. The point here is that Apple doesn't allow secret token to be valid more than 6 months so we need to renew the secret token sometimes. As an improvement we should renew the token only when expired. Is there a better way to store the private key ?

Reply:
We might have a button to regenerate the secret or something. Another approach, as I mentioned is re-using realm's keys so that we just reference the key in the config and generate the secret.

Reply:
Ok thanks, first I'll dig to find how to re-use the realm's key and then have a look to renew the secret when it has expired.

Reply:
But don't worry about realm's key too much. It was an idea and I'm not sure if it will fit here.

Btw, this method may also be useful `org.keycloak.broker.oidc.AbstractOAuth2IdentityProvider#getSignatureContext`. Also note that the base classes should also support JWT authentication already. See `org.keycloak.broker.oidc.AbstractOAuth2IdentityProvider#authenticateTokenRequest`.

Reply:
The signatureContext must conform the Apple spec (PKCS8 signature) and I was thinking `org.keycloak.broker.oidc.AbstractOAuth2IdentityProvider#getSignatureContext` was not based on the same signature, that is why I implemented my own signature method.

Reply:
Do you know if it is possible to update the stored configuration with the generated JWS to provide a key cache ?

Reply:
Maybe this org.keycloak.services.resources.admin.IdentityProviderResource#updateIdpFromRep can help. So we update only when the JWT is expired.

Not sure if the best approach though ... Need to think more about this ...